### ISSUE TITLE
Scatterplots in 2D grid layout do not have evenly spaced points

### ISSUE THREAD

**Issue by gokceneraslan** [matplotlib/matplotlib#13046]
### Bug report

**Bug summary**

When I try to plot a scatterplot where all points are positioned in a regular 2D-grid layout, plotted points are not evenly spaced. Spacings differ systematically which leads to strange regular vertical patterns. With a 100x100 grid, this is the case for (9, 9) or smaller figures. Patterns get less visible with larger figures and disappear around (12, 12)

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.tile(np.arange(100), 100)
y = np.repeat(np.arange(100), 100)

f, ax = plt.subplots(figsize=(7, 7))
ax.scatter(x, y, s=1, c='k')
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/1140359/50409159-5ef7f580-07be-11e9-9083-5ce18d34311a.png)

**Expected outcome**

Here is a figure of size (12, 12):

![image](https://user-images.githubusercontent.com/1140359/50409217-fa896600-07be-11e9-8ce9-24a8294472cd.png)

Here is a figure from R for comparison:

![image](https://user-images.githubusercontent.com/1140359/50409242-3de3d480-07bf-11e9-99f1-35cc481e0edf.png)

**Matplotlib version**

  * Operating system: osx 10.14
  * Matplotlib version: 3.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.6.6
  * Jupyter version (if applicable): 4.4.0

**Comment by jklymak** [comment]
This is pixel snapping which we do for efficiency reasons (as I understand it).  You won’t notice the issue as badly if you increase your dpi.  It is also not as big a deal in mplcairo, a third party backend, but it’s not in the library yet because it requires quite recent c compilers.

**Comment by gokceneraslan** [comment]
I'm using `InlineBackend.figure_format='retina'` now, do you mean it's even worse with regular figure format? Also, is there a way to turn the snapping off?

**Comment by gokceneraslan** [comment]
Maybe related to #7233?

**Comment by jklymak** [comment]
Yes snapping kwarg doesn’t do anything.    Again set your dpi to something higher when saving and it will be better.

**Comment by jklymak** [comment]
It is indeed a duplicate of #7233.  Let’s close this and continue discussion there if there are things to add. 

Again, I don’t think we need to do things this way, but it’s not trivial to change, and hence hasn’t been addressed except in newer backends.

**Comment by tacaswell** [comment]
The issue is not figure size (in inches) but figure size in pixels.

The issue is a fast-path optimization that draws the marker once and 'stamps' it onto the canvas where the markers should be.  This has the advantage of being faster in Agg, and results in smaller files for the vector backends, however the down side is that each marker is effectively rounded to the nearest pixel.  This case is perfectly designed to high-light the issue due to the regular, but incommensurate, grid.
